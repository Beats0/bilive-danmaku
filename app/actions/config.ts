import { Dispatch } from '../reducers/types';
import { getVersionInfo } from '../api';

export const UPDATECONFIG = 'UPDATECONFIG';
export const RESETCONFIG = 'RESETCONFIG';
export const FETCHVERSIONDATA = 'FETCHVERSIONDATA';

export interface UpdateConfigAction {
  type: typeof UPDATECONFIG;
  payload: UpdateConfigPayload;
}

export interface UpdateConfigPayload {
  k: string;
  v: number | string;
}

export interface ResetConfigAction {
  type: typeof RESETCONFIG;
}

export interface FetchVersionDataAction {
  type: typeof FETCHVERSIONDATA;
  payload: {
    latestVersion: string;
  };
}

export type ConfigActionTypes =
  | UpdateConfigAction
  | ResetConfigAction
  | FetchVersionDataAction;

// 更新配置
export function updateConfig(payload: UpdateConfigPayload) {
  return {
    type: UPDATECONFIG,
    payload
  };
}

// 恢复默认配置
export function resetConfig() {
  return {
    type: RESETCONFIG
  };
}

// 获取服务器端version
export function fetchVersionData() {
  return async (dispatch: Dispatch) => {
    const latestVersion = await getVersionInfo();
    const action: FetchVersionDataAction = {
      type: FETCHVERSIONDATA,
      payload: {
        latestVersion
      }
    };
    dispatch(action);
  };
}
