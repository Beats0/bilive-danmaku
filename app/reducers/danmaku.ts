import { danmakuStateType } from './types';
import Socket from '../components/Danmaku/base/Socket';
import {
  CREATESOCKET,
  FETCHGIFTDATA,
  DanmakuActionTypes
} from '../actions/danmaku';
import config from '../config';

const initState: danmakuStateType = {
  roomid: config.roomid,
  socket: new Socket(config.roomid),
  giftMap: new Map<number, GiftRaw>(),
  userAvatarMap: new Map<number, string>()
};

export default function danmaku(state = initState, action: DanmakuActionTypes) {
  switch (action.type) {
    case CREATESOCKET:
      const { roomid } = action.payload;
      return { ...state, roomid, socket: new Socket(roomid) };
    case FETCHGIFTDATA:
      const { giftMap } = action.payload;
      return { ...state, giftMap };
    default:
      return state;
  }
}
