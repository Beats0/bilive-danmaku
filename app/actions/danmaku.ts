import { Dispatch, GetState } from '../reducers/types';
import { getGiftInfo } from '../api';

export const CREATESOCKET = 'CREATESOCKET';
export const FETCHGIFTDATA = 'FETCHGIFTDATA';

export interface CreateSocketAction {
  type: typeof CREATESOCKET;
  payload: {
    roomid: number;
  };
}

export interface FetchGiftDataAction {
  type: typeof FETCHGIFTDATA;
  payload: {
    giftMap: Map<number, GiftRaw>;
  };
}

export type DanmakuActionTypes = CreateSocketAction | FetchGiftDataAction;

// 创建Socket
export function createSocket(roomid: number) {
  return (dispatch: Dispatch, getState: GetState) => {
    const { danmaku } = getState();
    // 关闭以前的socket
    danmaku.socket.close();
    const ation: CreateSocketAction = {
      type: CREATESOCKET,
      payload: {
        roomid
      }
    };
    dispatch(ation);
  };
}

// 获取礼物数据
export function fetchGiftData() {
  return async (dispatch: Dispatch) => {
    const giftMap = await getGiftInfo();
    const ation: FetchGiftDataAction = {
      type: FETCHGIFTDATA,
      payload: {
        giftMap
      }
    };
    dispatch(ation);
  };
}
