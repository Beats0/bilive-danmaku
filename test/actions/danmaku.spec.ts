import * as actions from '../../app/actions/danmaku';

describe('actions', () => {
  it('should increment should create init websocket action', () => {
    expect(actions.CREATESOCKET).toMatchSnapshot();
  });
  it('should increment should create fetch gift data api action', () => {
    expect(actions.FETCHGIFTDATA).toMatchSnapshot();
  });
});
