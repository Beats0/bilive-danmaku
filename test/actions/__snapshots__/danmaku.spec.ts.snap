// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should increment should create fetch gift data api action 1`] = `"FETCHGIFTDATA"`;

exports[`actions should increment should create init websocket action 1`] = `"CREATESOCKET"`;
